app/models/concerns/source_identifier_filterable.rb

```diff
@@ -0,0 +1,73 @@
+# frozen_string_literal: true
+
+# SourceIdentifierFilterable
+#
+# This module provides functionality for filtering records based on source identifiers.
+# A source identifier can be a plate barcode, tube barcode, or a combination of plate barcode
+# and well position.
+# Models that include this module will have methods to filter records using these identifiers.
+#
+# ## Methods:
+#
+# * apply_source_identifier_filter(records, value, plate_join: :plate,
+#                                  tube_join: :tube, well_join: :well)
+#   - Filters the given records based on the provided source identifiers.
+#   - Parameters:
+#     - records: The ActiveRecord relation to filter.
+#     - value: An array of source identifiers to filter by.
+#     - plate_join: The association name for joining with the plate table (default: :plate).
+#     - tube_join: The association name for joining with the tube table (default: :tube).
+#     - well_join: The association name for joining with the well table (default: :well).
+#
+# ## Example Usage:
+#
+#   class MyModel < ApplicationRecord
+#     include SourceIdentifierFilterable
+#   end
+#
+#   records = MyModel.all
+#   source_identifiers = ['PLATE123', 'TUBE456', 'PLATE789:A1']
+#   filtered_records = MyModel.apply_source_identifier_filter(records, source_identifiers)
+#
+#   # Custom join associations
+#   filtered_records = MyModel.apply_source_identifier_filter(records, source_identifiers,
+#                                                             plate_join: :source_plate,
+#                                                             tube_join: :source_tube)
+#
+module SourceIdentifierFilterable
+  extend ActiveSupport::Concern
+
+  class_methods do # rubocop:disable Metrics/BlockLength
```

Comment:
I am not sure it is the right idea to ignore this. This is a great piece of work but if you split it into functions it will be a lot more readable. Happy to go through it with you.

Reply:
Ok. I will see I can break it down further. 👍

Reply:
I suspect Rubocop would have some ideas too but because of the Metrics/BlockLength it won't make any more suggestions.

Reply:
I’ve broken the code into functions and modularized it using module functions. If these are changed to private class methods, RuboCop still raises concerns about the module size. Alternatively, wrapping these methods into another module and include it within this concern will allow  to remove 'block length' warning .

I’m not entirely convinced about the module size limits imposed by RuboCop  or the need to artificially modify the code just to satisfy it. While there is definitely value in splitting functions to improve readability and modularity, the imposed limit and the way RuboCop evaluates the structure feels bit unjustified.

Reply:
I understand your frustration and Rubocop can be dogmatic sometimes but looking at the result I find it far easier to understand what is going on. It is a question of maintainability. 

I think if we ignore Rubocop when we disagree with it we just end up with lots of stuff disabled which can defeat the purpose of it. I really struggle with long methods because I get overwhelmed and need to break it down. I would now feel far more confident modifying it because I know what is going on. The method names help me to understand. I hope that adds some context?

Reply:
Another thing this does for me is it helps me to identify places where we can make stuff more flexible. For example in this filter you have used plate, tube and well distinctions. I see this a lot in our code which leads me to believe that we need to do some work around labware. Having a big method can sometimes obfuscate this.

Reply:
Yes, absolutely. I fully agree with your point, and breaking down the structures absolutely made more sense. As you said, it is more readable, modular, and clear now. 
What I was referring to was the structure it seems to be approving of. It would have been equally good, or even better, to have the new methods added as private in-class methods, but RuboCop is not happy with that because of the size. This is the only point I am not convinced about.

> I understand your frustration and Rubocop can be dogmatic sometimes but looking at the result I find it far easier to understand what is going on. It is a question of maintainability.
> 
> I think if we ignore Rubocop when we disagree with it we just end up with lots of stuff disabled which can defeat the purpose of it. I really struggle with long methods because I get overwhelmed and need to break it down. I would now feel far more confident modifying it because I know what is going on. The method names help me to understand. I hope that adds some context?

Reply:
You are correct. It would have been better to have private in-class methods. Looking at what you have done I think there is a way we could reduce the size significantly by changing the structure. For example if we turned joins into a hash we could probably use map or reduce and build the filter in one go using Arel.

Reply:
I'm not entirely sure how to do it, may be we can do it by pairing on it at sometime later after finishing y24-149, if that's ok.

Reply:
Sure. I was going to have a crack but not quite sure either so pairing is a good idea.